feat: add C ndarray API and refactor `blas/ext/base/dsorthp`

0PrashantYadav0 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: na
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   add C ndarray API and refactor`blas/ext/base/dsorthp`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

0PrashantYadav0 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

0PrashantYadav0 committed:
chore: remove unwanted include

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
chore: updated manifest file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -48,9 +48,9 @@ The function has the following parameters:
+-   **strideX**: stride length.
 
-The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to sort every other element
+The `N` and stride parameters determine which elements in `x` are accessed at runtime. For example, to sort every other element
```
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to sort every other element:
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -92,9 +92,9 @@ dsorthp.ndarray( x.length, 1.0, x, 1, 0 );
+-   **offsetX**: starting index.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to access only the last three elements of `x`
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to access only the last three elements:
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on starting a index. For example, to access only the last three elements:
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
````diff
@@ -164,6 +149,126 @@ console.log( x );
+```c
+double x[] = { 1.0, -2.0, 3.0, -4.0 };
+
+stdlib_strided_dsorthp( 2, -1, x, 1 );
````
```suggestion
stdlib_strided_dsorthp( 2, -1.0, x, 1 );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
````diff
@@ -164,6 +149,126 @@ console.log( x );
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+
+```c
+stdlib_strided_dsorthp( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX );
````
```suggestion
stdlib_strided_dsorthp( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
````diff
@@ -164,6 +149,126 @@ console.log( x );
+```c
+double x[] = { 1.0, -2.0, 3.0, -4.0 };
+
+stdlib_strided_dsorthp_ndarray( 4, 1, x, 1, 0 );
````
```suggestion
stdlib_strided_dsorthp_ndarray( 4, 1.0, x, 1, 0 );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -164,6 +149,126 @@ console.log( x );
+The function accepts the following arguments:
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] CBLAS_INT` sort order.
```
```suggestion
-   **order**: `[in] double` sort order.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -164,6 +149,126 @@ console.log( x );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] CBLAS_INT` sort order.
+-   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
```
```suggestion
-   **X**: `[inout] double*` input array.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -164,6 +149,126 @@ console.log( x );
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] CBLAS_INT` sort order.
+-   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
```
```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -164,6 +149,126 @@ console.log( x );
+-   **order**: `[in] CBLAS_INT` sort order.
+-   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
```
```suggestion
-   **offsetX**: `[in] CBLAS_INT` starting index.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
````diff
@@ -164,6 +149,126 @@ console.log( x );
+-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
+
+```c
+stdlib_strided_dsorthp_ndarray( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX );
````
```suggestion
stdlib_strided_dsorthp_ndarray( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsorthp( len, 1, x, 1 );
```
```suggestion
		stdlib_strided_dsorthp( len, 1.0, x, 1 );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsorthp_ndarray( len, 1, x, 1, 0);
```
```suggestion
		stdlib_strided_dsorthp_ndarray( len, 1.0, x, 1, 0);
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/docs/repl.txt:
```diff
@@ -1,8 +1,8 @@
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
+    The `N` and stride parameters determine which elements in `x` are accessed
     at runtime.
```
```suggestion
    The `N` and stride parameters determine which elements in the strided array
    are accessed at runtime.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/docs/repl.txt:
```diff
@@ -56,25 +56,24 @@
     While typed array views mandate a view offset based on the underlying
-    buffer, the `offset` parameter supports indexing semantics based on a
+    buffer, the `offsetX` parameter supports indexing semantics based on a
     starting index.
```
```suggestion
    buffer, the offset parameter supports indexing semantics based on a starting
    index.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/docs/repl.txt:
```diff
@@ -89,10 +88,10 @@
 
-    offset: integer
+    offsetX: integer
         Starting index of `x`.
```
```suggestion
        Starting index.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/include/stdlib/blas/ext/base/dsorthp.h:
```diff
@@ -31,7 +31,12 @@ extern "C" {
 * Sorts a double-precision floating-point strided array using heapsort.
 */
-void c_dsorthp( const int64_t N, const double order, double *X, const int64_t stride );
+void API_SUFFIX(stdlib_strided_dsorthp)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX );
```
```suggestion
void API_SUFFIX(stdlib_strided_dsorthp)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/include/stdlib/blas/ext/base/dsorthp.h:
```diff
@@ -31,7 +31,12 @@ extern "C" {
+/**
+* Sorts a double-precision floating-point strided array using heapsort and alternative indexing semantics.
+*/
+void API_SUFFIX(stdlib_strided_dsorthp_ndarray)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX );
```
```suggestion
void API_SUFFIX(stdlib_strided_dsorthp_ndarray)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/lib/dsorthp.js:
```diff
@@ -53,96 +52,10 @@ var floor = require( '@stdlib/math/base/special/floor' );
+function dsorthp( N, order, x, strideX ) {
+	var offsetX;
+	offsetX = stride2offset( N, strideX );
+	return ndarray( N, order, x, strideX, offsetX );
```
```suggestion
	return ndarray( N, order, x, strideX, stride2offset( N, strideX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/src/main.c:
```diff
@@ -34,20 +35,41 @@
-	int64_t i;
-	int64_t j;
-	int64_t k;
+void API_SUFFIX(stdlib_strided_dsorthp)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dsorthp)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX) {
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/src/main.c:
```diff
@@ -34,20 +35,41 @@
+* @param strideX  stride length
+* @param offsetX  starting index
+*/
+void API_SUFFIX(stdlib_strided_dsorthp_ndarray)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX ) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dsorthp_ndarray)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```

headlessNode replied:
For this change you will also have to refactor the implementation below. You can take a look at #3292 as a reference.

headlessNode reviewed (commented):
@0PrashantYadav0 Thanks for working on this. Left an Initial round of review.

0PrashantYadav0 committed:
fix: update documentation and benchmarks to use double for order parameter

0PrashantYadav0 commented:
@headlessNode Thank you for review. I make the changes you asked for please review and merge this pr.

0PrashantYadav0 committed:
feat: add floor function dependency and refactor sorting logic in dsorthp

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
@headlessNode Thanks for review. I have make the changes you asked for please review them as well. Thank you.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md:
```diff
@@ -164,6 +149,126 @@ console.log( x );
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] double` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
+-   **X**: `[inout] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
```
```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsorthp_ndarray( len, 1.0, x, 1, 0);
```
```suggestion
		stdlib_strided_dsorthp_ndarray( len, 1.0, x, 1, 0 );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsorthp/src/main.c:
```diff
@@ -19,7 +19,9 @@
+#include "stdlib/blas/base/shared.h"
+#include "stdlib/strided/base/stride2offset.h"
+#include "stdlib/math/base/special/floor.h"
 #include <math.h>
```
```suggestion
```

headlessNode reviewed (changes requested):
Apart from these. LGTM.

0PrashantYadav0 committed:
fix: update README and benchmark for strideX parameter description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: failed
---
